lib/src/rewrite.rs

```diff
@@ -540,8 +562,18 @@ impl<'settings, 'repo> DescendantRebaser<'settings, 'repo> {
                 &new_parents,
                 &self.options,
             )?;
-            self.rebased
-                .insert(old_commit_id.clone(), new_commit.id().clone());
+            assert_eq!(
+                self.rebased
+                    .insert(old_commit_id.clone(), new_commit.id().clone()),
+                None,
+                "Trying to rebase the same commit in two different ways",
+            );
+            assert_eq!(
+                self.parent_mapping
+                    .insert(old_commit_id.clone(), vec![new_commit.id().clone()]),
+                None,
+                "Trying to rebase the same commit in two different ways",
+            );
```

Comment:
nit: I slightly prefer to move codes having side-effect out of the assertion, but I don't know if that's common practice.

Reply:
You mean `let result = self.parent_mapping.insert(...); assert(...)`?

I thought about it, but it felt a bit artificial. The only way I can think of that's really clear would be to assert on `self.parent_mapping.contains()` and then insert.

I can change it to whatever, though.

Reply:
> I slightly prefer to move codes having side-effect out of the assertion, but I don't know if that's common practice.

I agree. Rust's (non-`debug_`) assertions cannot be disabled, so maybe I only feel that way because of past experiences from other languages and there's no real reason to worry about it in Rust.

Reply:
I mainly inserted the asserts as documentation that would hopefully not go out of date. I'm happy to change the `debug_assert`-s, but that won't solve the problem of hiding a operation with a side-effect behind an assert.

We could remove the asserts entirely, but then any bugs leading from circular rebases might be a little harder to debug.

Currently, my best guess is that the following might work for everyone. WDYT?

```rust
debug_assert!(self.parent_mapping.contains(...));
self.parent_mapping.insert(...);
```

Ultimately, I'll go with whatever works for you.

Reply:
> Do you mean `let result = self.parent_mapping.insert(...); assert(...)`?

I would actually write that way (or `.insert().is_none(); assert!(...)`.) I think `debug_assert!(...contains())` is also good.

Reply:
OK, here's another attempt. Let me know if that looks OK.

Reply:
Looks better, thanks. It doesn't have to be *debug_* assertion since comparison should be cheap?

Reply:
> Rust's (non-debug_) assertions cannot be disabled

I just realized you meant this as a *good* thing; I misunderstood before.